Add Delayed Vetoable to deploy scripts

maurelian opened the pull request:
**Changes to `SuperchainConfig`:**

The `systemOwner` variable is now read directly from the `PROXY_OWNER_ADDRESS` slot in storage, rather than having a separate storage slot dedicated to it. Although this is breaking the separation of concerns between proxy and implementation, this change is important to avoid redundant config, and potential configuration errors. 

**Changes to `DelayedVetoable`**:

- Refactored `DelayedVetoable` to read from `SuperchainConfig`.
- Replaced immutable reads with internal getters.
- Added separate internal and external getters to properly support the `readOrHandle` modifier

Note: the goal of making `DelayedVetoable` an easily inheritable and reusable contract was abandoned, as it would have significantly impacted the maintainability and readability of the contract to keep both the pre-existing immutable vars, while overriding the internal methods to read from the `SuperchainConfig`.

**Changes to deployment scripts:**

- Deploys a Safe for the Initiator (security council) and Vetoer (foundation)
- Deploys a new `DelayedVetoable` in as admin of the `SuperchainConfig` 
- Deploys a new `DelayedVetoable` as owner of the `ProxyAdmin`
- Renamed config vars:
	- `finalSystemOwner` → `systemConfigOwner`
	- `portalGuardian` → `superchainConfigGuardian`

maurelian committed:
feat(ctb): Add virtual internal getters

This change breaks out the pre-existing external getters into pairs
of internal getters that read immutables, and external getters that
call the internal ones.
This is necessary to allow the contract to be inherited.

maurelian committed:
refactor(ctb): replace immutable reads with internal getters

maurelian committed:
refactor(ctb): Make DelayedVetoable read from SuperchainConfig

DelayedVetoable now reads most config values from SuperchainConfig.
The SuperchainConfig now treats the SystemOwner and Proxy owner
addresses as the same, which reduces config and avoids possible config
errors.

maurelian committed:
chore(ctb): Fix import paths in CommonTest.sol

maurelian committed:
fix(ctb): SuperchainConfig remove unused system owner refs

maurelian committed:
fix(ctb): L2OO removed unused SuperchainConfig import

maurelian committed:
feat(ctb): Deploy DelayedVetoable for SuperchainConfig

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -31,11 +30,6 @@ contract SuperchainConfig is Initializable, ISemver {
-    /// @notice The address of the systemOwner who may trigger an upgrade or change to critical config values.
-    ///         This will be a DelayedVetoable contract.
-    ///         It can only be modified by an upgrade.
-    bytes32 public constant SYSTEM_OWNER_SLOT = bytes32(uint256(keccak256("superchainConfig.systemowner")) - 1);
```
Hmm I don't love leaking the proxy admin abstraction into the implementation here

maurelian replied:
I 100% get where you're coming from with that, it felt a bit dirty to me too. I'm kind of on the fence, but in the interest of finding the best solution, here's my steel man argument for this approach:

The requirement is that both [upgrades](https://www.notion.so/oplabs/External-Security-Council-Product-Requirements-FAQ-10425231ed5a47138c388b8705c313f1?pvs=4#aebe5405146f48dd893ec4b344cb5401), and the [removal of a sequencer](https://www.notion.so/oplabs/External-Security-Council-Product-Requirements-FAQ-10425231ed5a47138c388b8705c313f1?pvs=4#98a95e1bb0b243d980d3b4edb4fbe9f2) must be  enacted by Optimism Governance, and subject to a delay.

The implementation can satisfy that requirement directly by reading the same address from storage, the alternative is to have:

- 2 separate values in the config which we need to assert are identical
- 2 separate storage values that need to be kept in sync

So yes, in one sense the proxy abstraction is leaking into the implementation. But in another sense, the intended behavior is being implemented directly, rather than enforced with additional checks.

maurelian replied:
I'm going to make sure this is documented in the PR description itself. Resolving.

maurelian committed:
feat(ctb): Deploy delayed vetoable on Op Chains

maurelian committed:
temp workaround on initiator/system owner

maurelian committed:
feat(ctb): Route ProxyAdmin calls via DelayedVetoable

maurelian committed:
chore: bind, lock and snap

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -416,6 +425,83 @@ contract Deploy is Deployer {
+    ///         Note: Because the getters on DelayedVetoable must be read by pranking as the zero
+    ///         address, this function does not use the broadcast modifier, and instead starts and
+    ///         stops broadcasting, then starts and stops pranking, within the function body.
+    function deployDelayedVetoableForSuperchain() public {
```
There is a `broadcast` modifier - does that not work because of the transparent nature based on `address(0)`?

maurelian replied:
Correct, you apparently can't prank during a broadcast.

maurelian committed:
refactor(config): Rename portalGuardian -> superchainConfigGuardian

maurelian committed:
refactor(config): Rename finalSystemOwner to systemConfigOwner

maurelian committed:
feat(ctb): Enable using any safe with _callViaSafe

maurelian committed:
feat(ctb): Replace system owner safe with initiator

Allows specifying own safe as the initiator, or deploying a new one.

maurelian committed:
feat(ctb): Activate delay at end of deployment.

maurelian committed:
feat(ctb): combine init functions into one

maurelian committed:
fix(ctb): typo delated

maurelian committed:
fix(chain-ops): use log.Warn correctly

tynes commented on packages/contracts-bedrock/deploy-config/getting-started.json:
```diff
@@ -1,6 +1,6 @@
-  "portalGuardian": "ADMIN",
+  "systemConfigOwner": "ADMIN",
+  "superchainConfigGuardian": "ADMIN",
 
```
We are going to need to add something to the op stack tutorial that says to check out an old commit or update it when this gets merged

maurelian replied:
Surfaced this in #7656.

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -71,45 +73,47 @@ contract Deploy is Deployer {
-        address supConfProxy = setupSuperchainConfig();
-        setupOpChain({ supConfProxy: supConfProxy, systemOwnerSafe: systemOwnerSafe });
+
+        address supConfInitiator;
```
This should be like:

```solidity
address supConfIgInitiator = cfg.superchainCOnfigInitiator();
if (supConfigInitiator == address(0) {
  supconfigInitator = deploySafe()
}
```

Otherwise supconfInitiator could be `address(0)` when passed to `setupOpChain`

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -191,7 +199,20 @@ contract Deploy is Deployer {
+        initializeProtocolVersions();
+    }
     // @notice Gets the address of the SafeProxyFactory and Safe singleton for use in deploying a new GnosisSafe.
+
```
extra newline?

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -209,7 +230,7 @@ contract Deploy is Deployer {
 
     /// @notice Deploy the Safe
-    function deploySafe() public broadcast returns (address addr_) {
+    function deploySafe(string memory safeName) public broadcast returns (address addr_) {
```
nit: prefixed arg

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -416,6 +438,89 @@ contract Deploy is Deployer {
+        vm.stopBroadcast();
+
+        vm.startPrank(address(0));
+        require(delayedVetoable.superchainConfig() == address(superchainConfigProxy));
```
Perhaps we could modularize the 2 deployments for delayed vetoable with an internal function to dedupe code

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -661,10 +774,11 @@ contract Deploy is Deployer {
                         (
-                            cfg.finalSystemOwner(), // systemOwner
-                            cfg.superchainConfigInitiator(), // initiator
+                            // todo: fix this
```
Wondering about this todo

maurelian replied:
My bad. 
The thing that needed to be fixed was the fact that if the initiator address is set in the config, then it won't be in the state when this script is run on a devnet. 

The fix is the code in `setupSuperchain` that deploys a new Safe if one is not provided.

I'll remove this todo.

tynes commented on packages/contracts-bedrock/src/L1/DelayedVetoable.sol:
```diff
@@ -78,39 +73,61 @@ contract DelayedVetoable is ISemver {
     string public constant version = "1.0.0";
```
nit: semver bump

tynes commented on packages/contracts-bedrock/src/L1/DelayedVetoable.sol:
```diff
@@ -42,17 +43,11 @@ contract DelayedVetoable is ISemver {
-    /// @notice The delay which will be set after the initial system deployment is completed.
-    uint256 internal immutable OPERATING_DELAY;
+    /// @notice The superchain config contract.
+    SuperchainConfig internal _superchainConfig;
```
any reason this has to be in storage?

maurelian replied:
No, good catch. It can be made immutable. I'll fix that.

Same goes for the `TARGET` var above. It looks like I put it into storage while moving the other immutables (`VETOER`, `INITIATATOR`, etc.) to the SuperchainConfig.

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -91,7 +85,6 @@ contract SuperchainConfig is Initializable, ISemver {
     /// @notice Constructs the SuperchainConfig contract.
```
nit: semver bump

maurelian replied:
`DelayedVetoable` needs a bump, yes, but this contract isn't on develop yet, so I think it should be OK at `1.0.0`, right?

maurelian replied:
bumped it anyways

maurelian commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -622,14 +725,22 @@ contract Deploy is Deployer {
+    )
+        internal
+    {
+        address delayedVetoableForOpChain = mustGetAddress("DelayedVetoableForOpChain");
```
I think this is broken. 
It shouldn't hard code the delayed vetoable contract to use. This'll require a minor refactor.

maurelian replied:
Fixed this.

tynes commented on packages/contracts-bedrock/src/L1/SuperchainConfig.sol:
```diff
@@ -237,13 +227,6 @@ contract SuperchainConfig is Initializable, ISemver {
-    /// @param _systemOwner The new system owner address.
-    function _setSystemOwner(address _systemOwner) internal {
-        Storage.setAddress(SYSTEM_OWNER_SLOT, _systemOwner);
-        emit ConfigUpdate(UpdateType.SYSTEM_OWNER, abi.encode(_systemOwner));
```
Interesting, now we lose this event, i think that is fine

maurelian replied:
As is the only place it was emitted was in the initializer.

tynes commented on op-chain-ops/genesis/config.go:
```diff
@@ -217,13 +215,19 @@ type DeployConfig struct {
 	SuperchainConfigSequencerKeyPairs []SequencerKeyPair `json:"superchainConfigSequencerKeyPairs"`
+	// Legacy config values.
+	// These are no longer used, but are necessary for parsing legacy deploy-configs
+	PortalGuardian   common.Address `json:"portalGuardian"`
```
We will want to remove these longer term

maurelian replied:
Agreed... I did not feel good about this, but I didn't really see another way available here.

tynes replied:
We should just delete these imo and bite the bullet

maurelian replied:
We agree to live with this for the moment.

maurelian committed:
feat(ctb): Add the ProtocolVersions contract to the Superchain

The ProtocolVersions contract is a singleton, so should be deployed in
superchainSetup. It also needed a new DelayedVetoable.

tynes commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -986,13 +1114,29 @@ contract Deploy is Deployer {
+            _target: mustGetAddress("DelayedVetoableForOpChain"),
+            _data: hex""
+        });
+        console.log("Delay activated on DelayedVetoable contracts");
```
What do you think about 2 loglines and include the addresses in the log?

tynes commented on packages/contracts-bedrock/src/L1/DelayedVetoable.sol:
```diff
@@ -42,17 +43,11 @@ contract DelayedVetoable is ISemver {
     /// @param data The data forwarded to the target.
```
Just a thought: this could be made into a clones-with-immutable-args factory + proxy pattern longer term, it would be cheaper to deploy minimal proxies that delegatecall a single impl, we shouldn't do this now but just something to think about

maurelian replied:
I have a strong feeling there is room for improvement wrt the operations of this contract.

tynes commented on packages/contracts-bedrock/test/CommonTest.t.sol:
```diff
@@ -97,11 +98,12 @@ contract SuperchainConfig_Initializer is CommonTest {
     address guardian = makeAddr("guardian");
-    uint256 delay = 100;
+    uint256 operatingDelay = 14 days;
+    // uint256 delay = operatingDelay;
```
nit: dead code

maurelian committed:
feat(ctb): Add resuable deployDelayedVetoable function

maurelian committed:
feat(ctb): Make DelayedVetoable storage immutable

maurelian committed:
chore(ctb): Bump semvers

maurelian committed:
fix(ctb): Address feedback

maurelian committed:
chore(ctb): Fix compiler warnings

tynes commented on op-chain-ops/genesis/config.go:
```diff
@@ -217,13 +215,19 @@ type DeployConfig struct {
+	// Legacy config values.
+	// These are no longer used, but are necessary for parsing legacy deploy-configs
+	PortalGuardian   common.Address `json:"portalGuardian"`
+	FinalSystemOwner common.Address `json:"finalSystemOwner"`
```
Thinking about this again, different values are used in the go code vs the solidity code, I am worried that if we leave both it will result in bugs. I would like to have these deduped before going to `develop`

maurelian replied:
I'm not sure I understand, but I've tracked this issue in the PR to `develop`. 

https://github.com/ethereum-optimism/optimism/pull/7656/files#r1358504501

tynes commented on packages/contracts-bedrock/test/CommonTest.t.sol:
```diff
@@ -125,14 +126,60 @@ contract SuperchainConfig_Initializer is CommonTest {
+    function setUp() public override {
+        super.setUp();
+        delayedVetoable = new DelayedVetoable({
+            superchainConfig_: supConf,
```
Why are these args postfixed instead of prefixed?

maurelian replied:
Probably autosuggested tbh.
Will fix, but it's crazy that we don't have linting for this.

maurelian replied:
OK, just for posterity, the reason for the postfix was that:
1. the internal getter function `_target` already had the prefix, so there was a naming collision.
2. I wanted to be consistent with the arg names on this function.

But I've resolved that differently by renaming it to `_targetContract`.

tynes commented on packages/contracts-bedrock/src/L1/DelayedVetoable.sol:
```diff
@@ -133,18 +150,18 @@ contract DelayedVetoable is ISemver {
     function _handleCall() internal {
         // The initiator and vetoer activate the delay by passing in null data.
```
nit: can cache calls onto the stack instead of making the same calls multiple times

maurelian replied:
Good call.
This pattern was safer and more iterative during refactoring, but caching finishes the job.

tynes commented on packages/contracts-bedrock/src/L1/DelayedVetoable.sol:
```diff
@@ -74,43 +69,65 @@ contract DelayedVetoable is ISemver {
-        INITIATOR = initiator_;
+    /// @param superchainConfig_ Address of the superchain config contract.
+    /// @param target_ Address of the target contract.
+    constructor(SuperchainConfig superchainConfig_, address target_) {
```
Why are these postfixed instead of prefixed?

tynes commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -420,7 +420,7 @@ func SystemConfig(batch *safe.Batch, implementations superchain.ImplementationLi
 	}
 
 	calldata, err := initialize.Inputs.PackValues([]any{
-		finalSystemOwner,
```
The base that this is implemented on has bugs in this file, should be a simple rebase when its ready

maurelian replied:
I'd like to do a rebase ASAP, appreciate the heads up.

maurelian committed:
fix(ctb): Rename DV constructor vars according to style guide

maurelian committed:
fix(ctb): Fix compiler warnings about shadowed var names
